be/src/vec/aggregate_functions/aggregate_function_window_funnel.h

```diff
@@ -72,138 +80,233 @@ WindowFunnelMode string_to_window_funnel_mode(const String& string) {
-                            events_timestamp[event_idx] = {first_timestamp, timestamp};
-                        }
-                        if (event_idx + 1 == max_event_level) {
-                            // Usually, max event level is small.
-                            return max_event_level;
+                    break;
+                }
+                const auto& event_column = block.get_by_position(column_idx);
+                const auto& event_data =
+                        assert_cast<const ColumnVector<UInt8>&>(*event_column.column).get_data();
+                match_row = simd::find_one(event_data.data(), row_count, match_row + 1);
+                if (match_row < row_count) {
+                    auto current_timestamp =
+                            binary_cast<NativeType, DateValueType>(timestamp_data[match_row]);
+                    bool is_matched = current_timestamp <= end_timestamp;
+                    if (is_matched) {
+                        if constexpr (WINDOW_FUNNEL_MODE == WindowFunnelMode::INCREASE) {
+                            is_matched = current_timestamp > prev_timestamp;
                         }
                     }
-                }
-            } else {
-                if (is_first_set && window_funnel_mode == WindowFunnelMode::FIXED) {
-                    for (size_t i = 0; i < events_timestamp.size(); i++) {
-                        if (!events_timestamp[i].has_value()) {
-                            return i;
+                    if (!is_matched) {
+                        break;
+                    }
+                    if constexpr (WINDOW_FUNNEL_MODE == WindowFunnelMode::INCREASE) {
+                        prev_timestamp =
+                                binary_cast<NativeType, DateValueType>(timestamp_data[match_row]);
+                    }
+                    if constexpr (WINDOW_FUNNEL_MODE == WindowFunnelMode::DEDUPLICATION) {
+                        bool is_dup = false;
+                        if (match_row != last_match_row + 1) {
+                            auto rows_to_match = match_row - last_match_row - 1;
+                            for (int tmp_column_idx = 1; tmp_column_idx < column_idx;
+                                 tmp_column_idx++) {
+                                const auto& tmp_event_column =
+                                        block.get_by_position(tmp_column_idx);
+                                const auto& tmp_event_data =
+                                        assert_cast<const ColumnVector<UInt8>&>(
+                                                *tmp_event_column.column)
+                                                .get_data();
+                                auto dup_match_row = simd::find_one(
+                                        tmp_event_data.data(), rows_to_match, last_match_row + 1);
+                                if (dup_match_row >= rows_to_match) {
```

Comment:
`dup_match_row >= rows_to_match` means not matched?

Reply:
fixed